### PR TITLE
[aes] Provide the initial key in two shares for security hardening

### PR DESCRIPTION
**Pull request by vogelpi**
This PR is related the masking of the cipher core (first-order masking = two-share implementation) and contains the part of the implementation that is NOT transparent to software and DV. I thus decided to factor that out into a separate PR. 

The PR contains a single commit that modifies the design such that the initial key has to be provided in two shares via two separate sets of CSRs (KEY_SHARE0 + KEY_SHARE1). The actual key used for encryption and passed to the currently unmasked cipher core is obtained by combining/XORing the two shares in hardware.

The `aes_test` executed on the Verilator model passes but I still need to adapt actual DV. Nevertheless I wanted to give especially software people the chance to have a look without getting lost in the actual masking PR of the cipher core.

### CODE REVIEW

**Comment by vogelpi** [comment]
Note that even if the cipher core itself wasn't masked, it can be beneficial to provide the key in two shares as this way the actual key is not exposed directly on the bus during configuration.

File: sw/device/tests/aes_test.c [hunk 1]
```diff
@@ -16,10 +16,21 @@ static const uint8_t plain_text_1[16] = {0x00, 0x11, 0x22, 0x33, 0x44, 0x55,
...
+
+// The masked key share (key XOR mask)
+static const uint8_t key_32_1_share0[32] = {
+    0x00 ^ 0x0f, 0x01 ^ 0x1f, 0x02 ^ 0x2f, 0x03 ^ 0x3f, 0x04 ^ 0x4f,
+    0x05 ^ 0x5f, 0x06 ^ 0x6f, 0x07 ^ 0x7f, 0x08 ^ 0x8f, 0x09 ^ 0x9f,
+    0x0a ^ 0xaf, 0x0b ^ 0xbf, 0x0c ^ 0xcf, 0x0d ^ 0xdf, 0x0e ^ 0xef,
+    0x0f ^ 0xff, 0x10 ^ 0x0a, 0x11 ^ 0x1a, 0x12 ^ 0x2a, 0x13 ^ 0x3a,
+    0x14 ^ 0x4a, 0x15 ^ 0x5a, 0x16 ^ 0x6a, 0x17 ^ 0x7a, 0x18 ^ 0x8a,
+    0x19 ^ 0x9a, 0x1a ^ 0xaa, 0x1b ^ 0xba, 0x1c ^ 0xca, 0x1d ^ 0xda,
+    0x1e ^ 0xea, 0x1f ^ 0xfa};
```
Comment (mcy) [inline sw/device/tests/aes_test.c:33]: Do you think this could be generated directly in the test?

Also, constants must be named `kConst`.

**Comment by sjgitty** [comment]
LG, would love to hear from @moidx and the security team to see how this aligns with our overall goals for hardware hardening.

File: sw/device/tests/aes_test.c (hunk 1 above)
Comment (vogelpi) [reply sw/device/tests/aes_test.c:33]: > Do you think this could be generated directly in the test?

I didn't want to give the impression that this is something software should do. Instead, software is supposed only ever see the two shares (and not the unmasked key + the mask). But maybe this could also be clarified with a comment.

> Also, constants must be named `kConst`.

I can do that.

**Comment by vogelpi** [comment]
I will give an update on that in today's security workgroup meeting.

File: hw/ip/aes/rtl/aes_core.sv [hunk 2]
```diff
@@ -243,6 +248,9 @@ module aes_core #(
   // Convert input data to state format (every input data word contains one state column)
   assign state_init = state_in ^ add_state_in;
 
+  // Combine the two key shares for the cipher core (currently not masked).
```
Comment (cdgori) [inline hw/ip/aes/rtl/aes_core.sv:251]: Maybe just indicate around this comment that this "will cause SCA leakage of the key" to be absolutely 100% clear?

File: hw/ip/aes/doc/_index.md [hunk 3]
```diff
@@ -313,16 +313,20 @@ This section discusses how software can interface with the AES unit.
 ## Initialization
 
 Before initialization, software must ensure that the AES unit is idle by checking {{< regref "STATUS.IDLE" >}}.
-If the AES unit is not idle, write operations to {{< regref "CTRL" >}}, the Initial Key registers {{< regref "KEY0" >}} - {{< regref "KEY7" >}} and initialization vector (IV) registers {{< regref "IV0" >}} - {{< regref "IV3" >}} are ignored.
+If the AES unit is not idle, write operations to {{< regref "CTRL" >}}, the Initial Key registers {{< regref "KEY_SHARE00" >}} - {{< regref "KEY_SHARE17" >}} and initialization vector (IV) registers {{< regref "IV0" >}} - {{< regref "IV3" >}} are ignored.
```
Comment (cdgori) [inline hw/ip/aes/doc/_index.md:316]: Is it "weird" for software that there is a "hole" between KEY_SHARE07 and KEY_SHARE10 ?  Should we make the names a little clearer (either SHAREA/SHAREB instead of SHARE0/SHARE1, or just adding some _ to the names)?  Or is this not needed? (I'm honestly not sure here.)

File: sw/device/lib/aes.c [hunk 4]
```diff
@@ -26,7 +26,8 @@ void aes_init(aes_cfg_t aes_cfg) {
   REG32(AES_CTRL_SHADOWED(0)) = cfg_val;
 };
 
-void aes_key_put(const void *key, aes_key_len_t key_len) {
+void aes_key_put(const void *key_share0, const void *key_share1,
```
Comment (cdgori) [inline sw/device/lib/aes.c:29]: We might need/want another function that takes a key that's not in shares, pulls in entropy/mask and puts mask in share0, key_input ^ mask in share1?

Sometimes we are given a key in raw form and need a way to feed it in.

File: sw/device/tests/aes_test.c (hunk 1 above)
Comment (vogelpi) [reply sw/device/tests/aes_test.c]: This is now changed.

File: hw/ip/aes/rtl/aes_core.sv (hunk 2 above)
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_core.sv]: Yep, that's done now.

File: hw/ip/aes/doc/_index.md (hunk 3 above)
Comment (vogelpi) [reply hw/ip/aes/doc/_index.md:316]: Thanks for raising this @cdgori . I agree with you that KEY_SHARE0_7 and KEY_SHARE1_7 would be better. I prefer the `_` instead of Share A and Share B. I have created another PR to modify the regtool for this here #2980 but it seems that needs to some more work.

File: sw/device/lib/aes.c (hunk 4 above)
Comment (vogelpi) [reply sw/device/lib/aes.c:29]: I think we really don't want to give the impression that this is something to do lightly. But I also see that this might be needed. Maybe this is something to keep in mind when updating the DIF? I am thus also pinging @silvestrst . But he is on vacation right now, so we might need to re-discuss this later.

Comment (lenary) [reply sw/device/lib/aes.c:29]: I think the DIF will always present a 2-share interface, and layers on top would have to manage the 1-share variant (because that requires interacting with other hardware, something DIFs are not allowed to do).

**Comment by vogelpi** [comment]
Update: Also adapted the DV infrastructure. DV sees the two separate shares rather than just transparently expanding the key to two shares when writing it to the registers via TL-UL. We probably need this to test corner case where just one share is updated and stuff like that. @rasmus-madsen , it would be great if you could have a look, too.

File: sw/device/lib/aes.c (hunk 4 above)
Comment (cdgori) [reply sw/device/lib/aes.c:29]: Ah good point about the interaction with the other hardware.  Your proposal for the layers on top to handle the 1-share is sensible.

**Comment by rasmus-madsen** [comment]
@vogelpi I would like to hold back on this until the issue I am currently seeing is fully debugged.
to make sure this does not mask it in anyway

File: hw/ip/aes/dv/env/seq_lib/aes_wake_up_vseq.sv [hunk 5]
```diff
@@ -32,7 +32,8 @@ class aes_wake_up_vseq extends aes_base_vseq;
     // set operation to encrypt
     set_operation(ENCRYPT);
 
-    `uvm_info(`gfn, $sformatf(" \n\t ---| WRITING INIT KEY  %02h", init_key), UVM_HIGH)
+    `uvm_info(`gfn, $sformatf(" \n\t ---| WRITING INIT KEY SHARE0 %02h", init_key[0]), UVM_HIGH)
```
[the remark this answers is not in the document]
Comment (vogelpi) [reply hw/ip/aes/dv/env/seq_lib/aes_wake_up_vseq.sv]: Done.

File: hw/ip/aes/dv/env/seq_lib/aes_wake_up_vseq.sv [hunk 6]
```diff
@@ -54,7 +55,8 @@ class aes_wake_up_vseq extends aes_base_vseq;
     // set aes to decrypt
     set_operation(DECRYPT);
     cfg.clk_rst_vif.wait_clks(20);
-    `uvm_info(`gfn, $sformatf("\n\t ---|WRITING INIT KEY FOR DECRYPT: %02h", init_key), UVM_HIGH)
+    `uvm_info(`gfn, $sformatf("\n\t ---|WRITING INIT KEY FOR DECRYPT SHARE0: %02h", init_key[0]), UVM_HIGH)
```
[the remark this answers is not in the document]
Comment (vogelpi) [reply hw/ip/aes/dv/env/seq_lib/aes_wake_up_vseq.sv]: Done.

**Comment by vogelpi** [comment]
> LGTM
> but we are seeing a few issues on AES that I think we should root cause first.
> so if possible delay the merge a day or two

Thanks for your feedback @rasmus-madsen . I agree, we should first fix the other issue. I am looking into it now.

**Comment by vogelpi** [comment]
Update: rebased on top of #3051, this still needs approval from a committer. Security (Chris) and DV (Rasmus) approved this PR already.

**Comment by vogelpi** [comment]
Thanks @tjaychen for approving this. Merging it finally :-)
